src/cmds/qdel.c

```diff
@@ -241,25 +327,23 @@ char **envp;
 			pbs_strncpy(warg, warg1, sizeof(warg));
 		}
 
-		stat = pbs_deljob(connect, job_id_out, warg);
+		p_delstatus = pbs_deljoblist(connect, pbs_svr->jobids, warg);
+		
+		while (p_delstatus != NULL) {
+			if ((pbse_to_txt(p_delstatus->code) != NULL) && (p_delstatus->code != PBSE_HISTJOBDELETED)) {
+				fprintf(stderr, "%s: %s %s\n", "qdel", pbse_to_txt(p_delstatus->code), p_delstatus->name);
+				any_failed = p_delstatus->code;
+			}
 
-		/*
-		 * The counter num_deleted should not be updated  when a history job is deleted .
-		 */
-		if (pbs_errno != PBSE_HISTJOBDELETED)
-			num_deleted++;
-		if (stat && (pbs_errno != PBSE_UNKJOBID && pbs_errno != PBSE_HISTJOBDELETED)) {
-			prt_job_err("qdel", connect, job_id_out);
-			any_failed = pbs_errno;
-		} else if (stat && (pbs_errno == PBSE_UNKJOBID) && !located) {
-			located = TRUE;
-			if (locate_job(job_id_out, server_out, rmt_server)) {
-				pbs_disconnect(connect);
-				pbs_strncpy(server_out, rmt_server, sizeof(server_out));
-				goto cnt;
+			if (p_delstatus->code != PBSE_HISTJOBDELETED)
+				num_deleted++;
+			
+			if (p_delstatus->code == PBSE_UNKJOBID) {
+				if (locate_job(p_delstatus->name, server_out, rmt_server))
+					add_jobid_to_pbs_servers(rmt_server, rmt_server, &pbs_servers);
```

Comment:
What will happen if rmt_server is server on which we already sent deljoblist request and finished its processing?

Reply:
Good thought, what you suggest for that corner case. ?

Reply:
I would look into moving locate_job() in req_deletejob on server side, as ultimately locate_job does pbs_connect() and then pbs_locjob() which goes to same server where we send deljob request, so same server can reply PBSE_UNKJOBID with located location as extra thing while replying to deljoblist request - this will save pbs_connect() (which is 1 req, 1 reply + all auth related part) and another pair of req/reply for pbs_locjob() - aka total 2 req/reply pair + auth stuff

Reply:
How this optimization would solve the question you have raised earlier? However, we find the moved destination server name, and that server was being processed already, how to send this deljoblist request to it. 

Whatever you are suggesting is applicable to many commands like, qsig, qstat, qhold, qmove, qmsg. Everywhere we are sending the request to the server to get the moved server name only if the return code is PBSE_UNKJOBID. Why now we need all the new optimization in this qdel PR. Do you feel is it worth to spend extra time to optimize this PR. It will simply cause us to delay the merge, I think the first priority should be to do the changes required to get the fastest qdel without breaking main functionalities, then later we can do all other innovative optimizations in pbs_server.

Reply:
@bremanandjk I think you misunderstand my point - Optimization is 2nd part, first I gave you suggestion for fixing issue which will come with this PR changes if we don't do anything.

> I would look into moving locate_job() in req_deletejob on server side, as ultimately locate_job does pbs_connect() and then pbs_locjob() which goes to same server where we send deljob request, so same server can reply PBSE_UNKJOBID with located location as extra thing while replying to deljoblist request

This is suggestion to fix the issue

>  this will save pbs_connect() (which is 1 req, 1 reply + all auth related part) and another pair of req/reply for pbs_locjob() - aka total 2 req/reply pair + auth stuff

This is optimization which will get free of cost if we do above suggestion.

Reply:
Another solution you can do is create another list of servers with list of located jobids (like you have pbs_servers now) then once all given jobids are processed you can process new servers list with located jobids. That way you won't have to move locate_job() code in req_deletejob().

Reply:
Right, this would be an apt solution. I am also implementing the same, Thanks